features/include/pcl/features/impl/good.hpp

```diff
@@ -0,0 +1,454 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2017-, Open Perception, Inc.
+ *
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder(s) nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifndef PCL_FEATURES_IMPL_GOOD_H_
+#define PCL_FEATURES_IMPL_GOOD_H_
+#include <pcl/features/good.h>
+#include <pcl/filters/project_inliers.h>
+#include <pcl/point_types.h>
+
+
+#include <numeric> ///std::accumulate
+
+//////////////////////////////////////////////////////////////////////////////////////////////  
+
+template <typename PointInT, int BinN> 
+void 
+pcl::GOODEstimation<PointInT, BinN>::computeBoundingBoxDimensions ()
+{
+  PointInT minimum_pt, maximum_pt, dimensions;  
+  pcl::getMinMax3D (*transformed_point_cloud_, minimum_pt, maximum_pt); // min max for bounding box
+  dimensions.getVector3fMap () = maximum_pt.getVector3fMap () - minimum_pt.getVector3fMap ();
+  bbox_dimensions_ << dimensions.x, dimensions.y, dimensions.z;
+}
+//////////////////////////////////////////////////////////////////////////////////////////////
+
+template <typename PointInT, int BinN>
+void
+pcl::GOODEstimation<PointInT, BinN>::projectPointCloudToPlane (PointCloudInPtr pc_in, pcl::ModelCoefficients::Ptr coefficients, PointCloudInPtr pc_out)
```

Comment:
- This method name is misleading. I propose to rename it to `fitPlaneToPointCloud`.
- `pc_in` needs to be `const PointCloudInConstPtr&`. You don't touch your input cloud and you don't want to trigger reference counting.
- `coefficients` needs to be `const pcl::ModelCoefficients::ConstPtr&`
- `pc_out` needs to be of type `PointCloud<PointInT>&`